src/Bicep.LangServer/Snippets/SnippetsProvider.cs

```diff
@@ -74,12 +84,163 @@ private void Initialize()
+            SemanticModel semanticModel = compilation.GetEntrypointSemanticModel();
+
+            return ResourceDependencyVisitor.GetResourceDependencies(semanticModel);
+        }
+
+        public Snippet GetResourceBodyCompletionSnippet(TypeSymbol typeSymbol)
+        {
+            string label = "{}";
+            string type = typeSymbol.Name;
+
+            StringBuilder sb = new StringBuilder();
+
+            // Get resource body completion snippet from checked in static template file, if available
+            if (resourceTypeToBodyMap.TryGetValue(type, out string? resourceBody))
+            {
+                sb.AppendLine(resourceBody);
+
+                if (resourceTypeToDependentsMap.TryGetValue(type, out string? resourceDependencies))
+                {
+                    sb.Append(resourceDependencies);
+                }
+
+                return new Snippet(sb.ToString(), CompletionPriority.Medium, label, label);
+            }
+
+            // Get resource body completion snippet from swagger spec
+            return GetResourceBodyCompletionSnippetFromSwaggerSpec(typeSymbol, label);
+        }
+
+        private Snippet GetResourceBodyCompletionSnippetFromSwaggerSpec(TypeSymbol typeSymbol, string label)
+        {
+            if (typeSymbol is ResourceType resourceType && resourceType.Body is ObjectType objectType)
+            {
+                int index = 1;
+                string? snippetText = null;
+
+                foreach (KeyValuePair<string, TypeProperty> kvp in objectType.Properties.OrderBy(x => x.Key))
+                {
+                    snippetText = GetSnippetText(kvp.Value, snippetText, ref index);
+                }
+
+                if (snippetText is not null)
+                {
+                    // Define final tab stop
+                    snippetText += "\t$0\n}";
+
+                    // Add to cache
+                    resourceTypeToBodyMap.Add(typeSymbol.Name, snippetText);
+
+                    return new Snippet(snippetText, CompletionPriority.Medium, label, label);
+                }
+            }
+
+            return new Snippet("{\n\t$0\n}", CompletionPriority.Medium, label, label);
+        }
+
+        private string? GetSnippetText(TypeProperty typeProperty, string? snippetText, ref int index)
+        {
+            if (typeProperty.Flags.HasFlag(TypePropertyFlags.Required))
+            {
+                if (snippetText is null)
+                {
+                    snippetText = "{\n";
+                }
+
+                if (typeProperty.TypeReference.Type is ObjectType objectType)
+                {
+                    snippetText += "\t" + typeProperty.Name + ": {\n";
+
+                    foreach (KeyValuePair<string, TypeProperty> kvp in objectType.Properties.OrderBy(x => x.Key))
+                    {
+                        snippetText = GetSnippetText(kvp.Value, snippetText, ref index);
```

Comment:
Should this be `+=`?

Reply:
Refactored the code here to use StringBuilder instead:
https://github.com/Azure/bicep/pull/2172/commits/4fc7f62bf3e8e4bba36946ab621b80b720207ef5